cmd/soroban-rpc/internal/db/event.go

```diff
@@ -0,0 +1,336 @@
+)
+
+// EventWriter is used during ingestion of events from LCM to DB
+type EventWriter interface {
+	InsertEvents(lcm xdr.LedgerCloseMeta) error
+}
+
+// EventReader has all the public methods to fetch events from DB
+type EventReader interface {
+	GetEvents(
+		ctx context.Context,
+		cursorRange CursorRange,
+		contractIDs [][]byte,
+		topics [][][]byte,
+		eventTypes []int,
+		f ScanFunction,
+	) error
+}
+
+type eventHandler struct {
+	log        *log.Entry
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+}
+
+func NewEventReader(log *log.Entry, db db.SessionInterface, passphrase string) EventReader {
+	return &eventHandler{log: log, db: db, passphrase: passphrase}
+}
+
+//nolint:gocognit,funlen,cyclop
+func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
+	txCount := lcm.CountTransactions()
+
+	if eventHandler.stmtCache == nil {
+		return errors.New("EventWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	var txReader *ingest.LedgerTransactionReader
+	txReader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(eventHandler.passphrase, lcm)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to open transaction reader for ledger %d: %w ",
+			lcm.LedgerSequence(), err)
+	}
+	defer func() {
+		closeErr := txReader.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	for {
+		var tx ingest.LedgerTransaction
+		tx, err = txReader.Read()
+		if errors.Is(err, io.EOF) {
+			err = nil
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if !tx.Result.Successful() {
+			continue
+		}
+
+		transactionHash := tx.Result.TransactionHash[:]
+
+		txEvents, err := tx.GetDiagnosticEvents()
+		if err != nil {
+			return err
+		}
+
+		if len(txEvents) == 0 {
+			continue
+		}
+
+		query := sq.Insert(eventTableName).
+			Columns(
+				"id",
+				"contract_id",
+				"event_type",
+				"event_data",
+				"ledger_close_time",
+				"transaction_hash",
+				"topic1", "topic2", "topic3", "topic4",
+			)
+
+		for index, e := range txEvents {
+			var contractID []byte
+			if e.Event.ContractId != nil {
+				contractID = e.Event.ContractId[:]
+			}
+
+			id := Cursor{Ledger: lcm.LedgerSequence(), Tx: tx.Index, Op: 0, Event: uint32(index)}.String()
+			eventBlob, err := e.MarshalBinary()
+			if err != nil {
+				return err
+			}
+
+			v0, ok := e.Event.Body.GetV0()
+			if !ok {
+				return errors.New("unknown event version")
+			}
+
+			// Encode the topics
+			topicList := make([][]byte, MaxTopicCount)
+			for index, segment := range v0.Topics {
+				seg, err := segment.MarshalBinary()
+				if err != nil {
+					return err
+				}
+				topicList[index] = seg
+			}
+
+			query = query.Values(
+				id,
+				contractID,
+				int(e.Event.Type),
+				eventBlob,
+				lcm.LedgerCloseTime(),
+				transactionHash,
+				topicList[0], topicList[1], topicList[2], topicList[3],
+			)
+		}
+
+		_, err = query.RunWith(eventHandler.stmtCache).Exec()
```

Comment:
What's the ignored return value here? Let's add a comment clarifying for future readers.

Reply:
it is last inserted ID. We dont care about this as it is `INSERT` query